protocol/xmrtaker/swap_state.go

```diff
@@ -636,6 +635,22 @@ func (s *swapState) ready() error {
 	}
 
 	if stage != contracts.StagePending {
+		if stage == contracts.StageReady {
+			log.Warnf("contract already set to ready, ignoring call to ready()")
+			return nil
+		}
+
+		if stage == contracts.StageCompleted {
+			log.Infof("contract aleady set to completed, ignoring call to ready() and sending EventExit")
+			go func() {
+				err = s.Exit()
+				if err != nil {
+					log.Errorf("failed to handle EventExit: %s", err)
+				}
+			}()
```

Comment:
Same question as the other PR. What is the background goroutine for? Exit takes a long time, fair locking, something else? Is the code safe if the entire go routine completes before we return from the outer function here?

Reply:
it's because this code is called from inside the event handler, so it causes a deadlock if we try to call `Exit()` as it'll try to put something in to the event channel, but the code path we're in is on a read from the event channel. the other solution would be to make the event channel buffered, as it only has capacity 1 right now, but I'm not sure how ideal that is. can try in another PR

Reply:
In a follow-up PR, if we don't make it buffered, we should add code comments describing what you wrote above. That is subtle if you have to figure it out just by looking at the code.